lib/attrsets.nix

```
@@ -644,8 +644,10 @@ rec {
   filterAttrs =
     pred:
     set:
-    listToAttrs (concatMap (name: let v = set.${name}; in if pred name v then [(nameValuePair name v)] else []) (attrNames set));
-
+    let
+      attrsToRemove = lib.concatMap (name: if pred name set.${name} then [ ] else [ name ]) (attrNames set);
+    in
+    builtins.removeAttrs set attrsToRemove;
```

```suggestion
    removeAttrs set (concatMap (name: if pred name set.${name} then [ ] else [ name ]) (attrNames set));
```

Avoids an additional thunk and some selections :)

You'll need to add `removeAttrs` to the `inherit (builtins)` at the top of this file though.